specutils/spectra/spectral_axis.py

```diff
@@ -0,0 +1,95 @@
+import warnings
+
+import astropy.units as u
+from astropy.utils.decorators import lazyproperty
+import numpy as np
+
+from .spectral_coordinate import SpectralCoord
+
+
+__all__ = ['SpectralAxis']
+
+# We don't want to run doctests in the docstrings we inherit from Quantity
+__doctest_skip__ = ['SpectralAxis.*']
+
+class SpectralAxis(SpectralCoord):
+    """
+    Coordinate object representing spectral values corresponding to a specific
+    spectrum. Overloads SpectralCoord with additional information (currently
+    only bin edges).
+
+    Parameters
+    ----------
+    value : ndarray or `~astropy.units.Quantity` or `SpectralCoord`
+        Spectral axis data values.
+    unit : str or `~astropy.units.Unit`
+        Unit for the given data.
+    observer : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`, optional
+        The coordinate (position and velocity) of observer.
+    target : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`, optional
+        The coordinate (position and velocity) of observer.
+    radial_velocity : `~astropy.units.Quantity`, optional
+        The radial velocity of the target with respect to the observer.
+    redshift : float, optional
+        The redshift of the target with respect to the observer.
+    doppler_rest : `~astropy.units.Quantity`, optional
+        The rest value to use for velocity space transformations.
+    doppler_convention : str, optional
+        The convention to use when converting the spectral data to/from
+        velocity space.
+    bin_specificatoin: str, optional
+        Must be "edges" or "centers". Determines whether specified axis values
+        are interpreted as bin edges or bin centers. Defaults to "centers".
+    """
+
+    def __new__(cls, value, *args, bin_specification="centers", **kwargs):
+
+        # Convert to bin centers if bin edges were given, since SpectralCoord
+        # only accepts centers
+        if bin_specification == "edges":
+            bin_edges = value
+            value = SpectralAxis._centers_from_edges(value)
+
+        obj = super().__new__(cls, value, *args, **kwargs)
+
+        if bin_specification == "edges":
+            obj._bin_edges = bin_edges
+
+        return obj
+
+    def __quantity_subclass__(self, unit):
+        """
+        Overridden by subclasses to change what kind of view is
+        created based on the output unit of an operation.
+        """
+        return SpectralAxis, True
+
+    @staticmethod
```

Comment:
Just out of curiosity, is it possible that use cases exist where a 'bin center' is not the average of its edges? I am thinking of a (perhaps hypothetical) situation where a combination of a very non-linear dispersion curve with wide detector bins would justify such asymmetric bins. In that case, the code below this line would not work properly. I have in mind such things as the ACS grisms, where at the red end the dispersion becomes highly non-linear.

Reply:
I think the above is a reasonable first assumption, and we ought to state this in the documentation. Future work may address customizable edge -> bin calculations, or there could be a diff check on spectral axis array and if they aren't consistent bin sizes, raise a warning to the user that they should pre-calculate the centers and feed those in instead.